api/client.go

```diff
@@ -122,6 +122,7 @@ func (c *httpClient) Do(ctx context.Context, req *http.Request) (*http.Response,
 	if ctx != nil {
 		req = req.WithContext(ctx)
 	}
+	defer c.client.CloseIdleConnections()
```

Comment:
Are you sure this is a correct solution? 

My understanding is that this closes connections which might be REUSED on the subsequent `Do` request. So you have to establish them again (overhead)?

Source: https://forum.golangbridge.org/t/when-should-i-use-client-closeidleconnections/19254, https://stackoverflow.com/questions/18697290/apache-httpclient-how-to-auto-close-connections-by-servers-keep-alive-time

Reply:
Currently, doing an HTTP probe requiring allocating a new http.Client and http.Transport, and the underlying connections managed by http.Transport might be ESTABLISHED indefinitely if no parties (client or server) directly close them. As a result, if the probes continuously happen, we can see TCP connections opened by one process
drastically increase.

I understand that there may be such a problem in high concurrency. Even if it is reused, the connection is limited and cannot meet the requirements. In this case, the connection may be occupied and new connections cannot be applied for. As a result, the connection request cannot be continued. 

PS. I appeared in a high-concurrency scenario, with about 10000  of concurrencies.

Reply:
Nice! What "HTTP probe" do you mean? Is it some OSS code or your own app?

Generally it sounds you might use this client incorrectly if you create new client for single Do, assuming you can share the client.

For cases you cannot share e.g. it's serverless or it's on-off request to adhoc destination, I think you can access [Config.Client](https://github.com/prometheus/client_golang/blob/e74129ab0f4c33b6072bf62b5e3bf550371b5f5d/api/client.go#L46) (which is used to create API client) and invoke this method after `Do` on your own, which would be a correct thing to do 👍🏽 

I am also open to create `**CloseIdleConnections()**` method on our client with a clear commentary when it's useful to make it easier (: 

Doing this on every Do for *everybody* is incorrect.

Reply:
in this conde add Close func ? 
```
// Client is the interface for an API client.
type Client interface {
	URL(ep string, args map[string]string) *url.URL
	Do(context.Context, *http.Request) (*http.Response, []byte, error)
}
```

I think it would be more reasonable if the interface or open functions could be adjusted and allowed to be called by users themselves. Very useful to me too.

Reply:
@bwplotka  i am add code to interface  using CloseIdleConnections() func . plz review this code ?

Reply:
Why do we need interface change? Can we remove this defer? Otherwise new method is good 👍🏽

Reply:
Ok interface change is needed. Let's kill the defer statement here and LGTM!